File: day1/frequency.clj

(require '[clojure.java.io :as io])

(defn read-input
  "Read input filename and return a sequence of integers"
  [filename]
  (with-open [rdr (io/reader filename)]
    (doall (map #(Integer. %) (line-seq rdr)))))

(defn compute-frequency
  "Sum all numbers from input file"
  [filename]
  (reduce + (read-input filename)))

(defn find-repeats
  "Find first frequency that repeats"
  [filename]
  (let [input (into [] (read-input filename))
        length (count input)]
    (loop [idx 0
           current 0
           seen (set '(0))]
      (let [freq (+ (nth input (mod idx length)) current)]
        (if (contains? seen freq)
          freq
          (recur (+ idx 1)
                 freq
                 (conj seen freq)))))))

File: day1/frequency.py

def compute_frequency(input_fn):
    with open(input_fn, 'r') as fd:
        current_frequency = sum(int(freq) for freq in fd)
    return current_frequency

def find_repeats(input_fn):
    with open(input_fn, 'r') as fd:
        freqs = [int(freq) for freq in fd]

    seen = {0}
    current = 0

    while True:
        for freq in freqs:
            current += freq
            if current in seen:
                return current
            seen.add(current)

    return

def main():
    #print(compute_frequency('input.txt'))
    print(find_repeats('input.txt'))
    
if __name__ == '__main__':
    main()
    
